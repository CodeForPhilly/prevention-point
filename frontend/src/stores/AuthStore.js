import { observable, action, flow } from "mobx"
import { createContext } from "react"
import api from "../api"
import { handleError } from "../error"

export class AuthStore {
  constructor(rootStore) {
    this.rootStore = rootStore
  }

  @observable isAuthenticated = null
  @observable username = null
  @observable email = null
  @observable error = null

  @action
  setIsAuthenticated(auth) {
    this.isAuthenticated = auth
  }
  @action
  setUsername(username) {
    this.username = username
  }
  @action
  setEmail(email) {
    this.username = email
  }
  @action
  setError(error) {
    this.error = error
  }
  @action
  logout() {
    document.cookie = "JWT_ACCESS=; expires=Thu, 01 Jan 1970 00:00:01 GMT;"
    document.cookie = "JWT_REFRESH=; expires=Thu, 01 Jan 1970 00:00:01 GMT;"
    this.isAuthenticated = false
    this.username = null
    this.email = null
  }

  login = flow(function*(username, password) {
    try {
      const { ok, data, status } = yield api.createToken(username, password)
      if (!ok) {
        throw new Error(status)
      }
      this.setIsAuthenticated(true)
      this.setUsername(data.username)
      this.setEmail(data.email)
      this.setError(data.null)
    } catch (error) {
      const errorMessage = handleError(error.message)
      throw errorMessage
    }
  })
}

export const AuthStoreContext = createContext(new AuthStore())
