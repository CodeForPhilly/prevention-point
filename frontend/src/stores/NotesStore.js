import { observable, action, flow } from "mobx"
import { createContext } from "react"
import api from "../api"

export class NotesStore {
  constructor(rootStore) {
    this.rootStore = rootStore
  }

  @observable notesList = null

  @action
  setNotesList(note) {
    this.notesList = this.notesList.push(note)
  }

  visitsList = flow(function*(req) {
    try {
      const { ok, data } = yield api.getVisits(req)
      if (ok) {
        this.setNotesList(data)
      }
    } catch (error) {
      // TODO: Handle errors
    }
  })
}

export const NotesStoreContext = createContext(new NotesStore())
