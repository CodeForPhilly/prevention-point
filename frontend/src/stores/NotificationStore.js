import { observable, action } from "mobx"
import { createContext } from "react"

export class NotificationStore {
  constructor(rootStore) {
    this.rootStore = rootStore
  }

  // snackbar notifications
  @observable snackbarState = { message: "", open: false }

  @action setSnackbarState = (message, options = { open: true }) => {
    // we setup otions like this to maybe add more later
    this.snackbarState.message = message
    this.snackbarState.open = options.open
  }
}

export const NotificationStoreContext = createContext(new NotificationStore())
