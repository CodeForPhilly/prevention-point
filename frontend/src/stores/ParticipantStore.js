import { observable, action, flow } from "mobx"
import participantApi from "../api/participantApi"

class ParticipantStore {
  constructor(rootStore) {
    this.rootStore = rootStore
    this.getParticipants()
  }

  //participants = observable([])
  @observable participants = []

  @observable todos = ["buy milk", "buy eggs"]

  @action setParticipant(participant, index) {
    this.participants[index] = participant
  }

  getParticipants = flow(function*() {
    try {
      const data = yield participantApi.getParticipants()
      if (data) {
        data.forEach((datum, index) => {
          this.setParticipant(datum, index)
        })
      }
    } catch (error) {
      // TODO: Handle errors
    }
  })
}

let participantStore = (window.participantStore = new ParticipantStore())
export default participantStore
// uncomment this line to have the store on the dom and testable
// const store = (window.store = new ParticipantStore())
// store.getParticipants()
// export const participantStore = store

//export const participantStoreContext = createContext(new ParticipantStore())
//export const participantStoreContext = createContext(store)
