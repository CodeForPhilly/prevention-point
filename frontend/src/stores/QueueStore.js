import { observable, action, flow } from "mobx"
//import { createContext } from "react"
import api from "../api"

export class QueueStore {
  constructor(rootStore) {
    this.rootStore = rootStore
  }

  queueIds = {
    stepQueue: 1,
    legalServiceQueue: 2,
    needleExchangeQueue: 3,
  }

  @observable stepQueue = []
  @observable legalServiceQueue = []
  @observable needleExchangeQueue = []

  @action
  setQueue(queue, data) {
    queue = data
  }

  updateQueue = flow(function*(queue) {
    const { ok, data } = yield api.getQueue(this.queueIds["queue"])
    if (ok) {
      this.setQueue(queue, data)
    } else {
      // TODO: Handle errors
    }
  })
}
//export const QueueStoreContext = createContext(new QueueStore())
export const queueStore = new QueueStore()
