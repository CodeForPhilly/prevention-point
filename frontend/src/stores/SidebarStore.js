import { observable, decorate } from "mobx"
import { createContext } from "react"

export class SidebarStore {
  constructor(rootStore) {
    this.rootStore = rootStore
  }

  sidebarState = false

  drawerWidth = 400
}

decorate(SidebarStore, {
  sidebarState: observable,
  drawerWidth: observable,
})

export const SidebarStoreContext = createContext(new SidebarStore())
