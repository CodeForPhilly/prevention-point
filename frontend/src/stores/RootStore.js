import { createContext } from "react"
import { AuthStore } from "./AuthStore"
import { ParticipantStore } from "./ParticipantStore"
import { QueueStore } from "./QueueStore"

export class RootStore {
  // If creating a new store dont forget to add it here.
  authStore = new AuthStore(this)
  ParticipantStore = new ParticipantStore(this)
  QueueStore = new QueueStore(this)
}

export const rootStoreContext = createContext(new RootStore())
