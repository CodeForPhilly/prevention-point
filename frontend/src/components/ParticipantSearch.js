import React, { useContext } from "react"
import { observer } from "mobx-react-lite"
import { useHistory } from "react-router-dom"
import { makeStyles } from "@material-ui/core/styles"
import Grid from "@material-ui/core/Grid"
import Container from "@material-ui/core/Container"
import InputLabel from "@material-ui/core/InputLabel"
import FormControl from "@material-ui/core/FormControl"
import PrevPointInput from "./Input/PrevPointInput"
import PrevPointButton from "./PrevPointButton"
import PrevPointCopy from "./Typography/PrevPointCopy"
import { rootStoreContext } from "../stores/RootStore"
import PrevPointHeading from "./Typography/PrevPointHeading"
import { Formik, Form } from "formik"
import * as Yup from "yup"

const useStyles = makeStyles({
  root: {
    paddingTop: "20px",
    "& .error": {
      borderBottom: "1px solid red",
    },
  },
  form: {
    margin: "0 auto",
    maxWidth: "270px",
  },
  errorMessage: {
    color: "red",
  },
  heading: {
    padding: "2px 0",
    color: "#086375",
  },
  copy: {
    paddingTop: "20px",
  },
})

const ParticipantSearch = observer(() => {
  const classes = useStyles()
  const rootStore = useContext(rootStoreContext)
  const participantStore = rootStore.ParticipantStore
  const history = useHistory()

  const errorMessage = "Please enter a participant id or a name"
  const validationSchema = Yup.object().shape(
    {
      pp_id: Yup.string().when(["first_name", "last_name"], {
        is: (firstName, lastName) => !firstName && !lastName,
        then: Yup.string().required(errorMessage),
      }),
      first_name: Yup.string().when(["pp_id", "last_name"], {
        is: (ppId, lastName) => !ppId && !lastName,
        then: Yup.string().required(errorMessage),
      }),
      last_name: Yup.string().when(["pp_id", "first_name"], {
        is: (ppId, firstName) => !ppId && !firstName,
        then: Yup.string().required(errorMessage),
      }),
    },
    [
      ["pp_id", "first_name"],
      ["pp_id", "last_name"],
      ["first_name", "last_name"],
    ]
  )

  return (
    <Container className={classes.root}>
      <Formik
        className={classes.form}
        validateOnChange={false}
        validateOnBlur={false}
        initialValues={{
          pp_id: participantStore.params.pp_id
            ? participantStore.params.pp_id
            : "",
          first_name: participantStore.params.firstName
            ? participantStore.params.firstName
            : "",
          last_name: participantStore.params.lastName
            ? participantStore.params.lastName
            : "",
        }}
        validationSchema={validationSchema}
        onSubmit={(values, { setSubmitting }) => {
          participantStore.setParticipantIdParam(values.pp_id)
          participantStore.setParticipantFirstNameParam(values.first_name)
          participantStore.setParticipantLastNameParam(values.last_name)
          participantStore.getParticipants()
          history.push("/participants")
          setSubmitting(false)
        }}
      >
        {({ errors, touched, values, handleChange, isSubmitting }) => (
          <Form>
            <Grid item xs={12}>
              <PrevPointHeading className={classes.heading}>
                Participant Search
              </PrevPointHeading>
              <PrevPointCopy className={classes.copy}>
                <b>Reminder:</b> Search for participant profile prior to
                creating a new profile
              </PrevPointCopy>
            </Grid>
            <Grid item xs={12}>
              <FormControl error={errors.pp_id && touched.pp_id}>
                <InputLabel htmlFor="participant_id">Participant ID</InputLabel>
                <PrevPointInput
                  id="participant_id"
                  name="pp_id"
                  value={values.pp_id}
                  onChange={handleChange}
                />
              </FormControl>
            </Grid>
            <Grid item xs={12}>
              <PrevPointHeading className={classes.heading}>
                Or
              </PrevPointHeading>
            </Grid>
            <Grid item xs={12}>
              <FormControl error={errors.first_name && touched.first_name}>
                <InputLabel htmlFor="first_name">First Name</InputLabel>
                <PrevPointInput
                  id="first_name"
                  name="first_name"
                  value={values.firstName}
                  onChange={handleChange}
                />
              </FormControl>
            </Grid>
            <Grid item xs={12}>
              <FormControl error={errors.last_name && touched.last_name}>
                <InputLabel htmlFor="last_name">Last Name</InputLabel>
                <PrevPointInput
                  id="last_name"
                  name="last_name"
                  value={values.lastName}
                  onChange={handleChange}
                />
              </FormControl>
            </Grid>
            <Grid item xs={12}>
              <PrevPointButton type="submit" disabled={isSubmitting}>
                Submit
              </PrevPointButton>
              {(errors.pp_id || errors.first_name || errors.last_name) && (
                <PrevPointCopy className={classes.errorMessage}>
                  {errors.pp_id || errors.first_name || errors.last_name}
                </PrevPointCopy>
              )}
            </Grid>
          </Form>
        )}
      </Formik>
    </Container>
  )
})

export default ParticipantSearch
