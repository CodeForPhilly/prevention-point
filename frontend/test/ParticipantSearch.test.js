/* eslint-disable quotes */
import React from "react"
import { configure } from "enzyme"
import { createMount } from "@material-ui/core/test-utils"
import Adapter from "enzyme-adapter-react-16"
import { ThemeProvider, createMuiTheme } from "@material-ui/core/styles"

import ParticipantSearch from "../src/components/ParticipantSearch"
import { RootStoreContext, RootStore } from "../src/stores/RootStore"

configure({ adapter: new Adapter() })

const theme = createMuiTheme({})

describe("<ParticipantSearch />", () => {
  let wrapper
  let mount

  const initialProps = {}

  beforeAll(() => {
    mount = createMount()
    wrapper = mount(
      <RootStoreContext.Provider value={new RootStore()}>
        <ThemeProvider theme={theme}>
          <ParticipantSearch {...initialProps} />
        </ThemeProvider>
      </RootStoreContext.Provider>
    )
  })

  // what to do after each test
  afterAll(() => {
    jest.clearAllMocks()
  })

  it("should work", () => {
    wrapper = mount(
      <RootStoreContext.Provider value={new RootStore()}>
        <ParticipantSearch />
      </RootStoreContext.Provider>
    )
  })

  it("should match the snapshot", () => {
    expect(wrapper.html()).toMatchSnapshot()
  })

  it("should have a participants id", () => {
    expect(wrapper.exists('[name="pp_id"]')).toEqual(true)
  })

  it("should have a first name", () => {
    expect(wrapper.exists('[name="first_name"]')).toEqual(true)
  })

  it("should have a last name", () => {
    expect(wrapper.exists('[name="last_name"]')).toEqual(true)
  })

  it("should have a submit button", () => {
    expect(wrapper.exists("submit"))
  })

  // UI Integrity test
  it("should have submit button click", () => {
    wrapper
      .find("button")
      .at(0)
      .simulate("click")
  })
})
// resolves useHistory error hangup
jest.mock("react-router-dom", () => ({
  useHistory: () => ({
    push: jest.fn(),
  }),
}))
