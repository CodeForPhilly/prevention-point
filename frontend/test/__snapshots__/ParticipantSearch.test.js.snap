// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ParticipantSearch /> should match the snapshot 1`] = `"<div class=\\"MuiContainer-root makeStyles-root-7 MuiContainer-maxWidthLg\\"><form action=\\"#\\"><div class=\\"MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12\\"><h2 class=\\"MuiTypography-root makeStyles-heading-10 MuiTypography-h2 MuiTypography-displayBlock\\" aria-label=\\"heading\\">Participant Search</h2><p class=\\"MuiTypography-root makeStyles-copy-11 MuiTypography-body1 MuiTypography-displayBlock\\"><b>Reminder:</b> Search for participant profile prior to creating a new profile</p></div><div class=\\"MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12\\"><div class=\\"MuiFormControl-root\\"><label class=\\"MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated\\" data-shrink=\\"false\\" for=\\"participant_id\\">Participant ID</label><div class=\\"MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl\\"><input aria-invalid=\\"false\\" id=\\"participant_id\\" name=\\"pp_id\\" type=\\"text\\" class=\\"MuiInputBase-input MuiInput-input\\" value=\\"\\"></div></div></div><div class=\\"MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12\\"><h2 class=\\"MuiTypography-root makeStyles-heading-10 MuiTypography-h2 MuiTypography-displayBlock\\" aria-label=\\"heading\\">Or</h2></div><div class=\\"MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12\\"><div class=\\"MuiFormControl-root\\"><label class=\\"MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated\\" data-shrink=\\"false\\" for=\\"first_name\\">First Name</label><div class=\\"MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl\\"><input aria-invalid=\\"false\\" id=\\"first_name\\" name=\\"first_name\\" type=\\"text\\" class=\\"MuiInputBase-input MuiInput-input\\" value=\\"\\"></div></div></div><div class=\\"MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12\\"><div class=\\"MuiFormControl-root\\"><label class=\\"MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated\\" data-shrink=\\"false\\" for=\\"last_name\\">Last Name</label><div class=\\"MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl\\"><input aria-invalid=\\"false\\" id=\\"last_name\\" name=\\"last_name\\" type=\\"text\\" class=\\"MuiInputBase-input MuiInput-input\\" value=\\"\\"></div></div></div><div class=\\"MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12\\"><button class=\\"MuiButtonBase-root MuiButton-root MuiButton-contained makeStyles-PrevPointButton-12 MuiButton-containedPrimary MuiButton-containedSizeSmall MuiButton-sizeSmall\\" tabindex=\\"0\\" type=\\"submit\\"><span class=\\"MuiButton-label\\">Submit</span><span class=\\"MuiTouchRipple-root\\"></span></button></div></form></div>"`;
